### ISSUE TITLE
Utilizing Same API Client Pattern in another API Client Library

### ISSUE THREAD

**Issue by jon-whit** [google/go-github#915]
First off, major kudos to the authors of this API client library. It's really well written and uses some great design patterns for API client libraries. I'm super impressed!

That being said, I'm also working on an API client library that interacts with another API that is *not* GitHub. Is it ok if I reference similar structures and design patterns that I've noticed in this source in my own API client library?

I wanted to make sure I asked for approval of this, because the BSD license isn't necessarily clear in that regard.

**Comment by gmlewis** [comment]
@willnorris is the original author of this library and deserves the lion's share of your compliments. Thank you for the kind words! Will is also the person who can give you guidance on derivative works.

**Comment by jon-whit** [comment]
@gmlewis and @shurcooL thank you and keep up the awesome work!

**Comment by willnorris** [comment]
Thanks, I'm glad it's been useful for you.  

I can't give you specific legal advice on open source licenses, but I can give you some general principles:

- if you're actually copying and pasting code from an open source library, then the original license of the library generally still applies.  All open source allow for these kinds of derivative works, though some impose certain additional obligations.  BSD is just a "notice style" license, which only requires acknowledgement.
- ideas aren't subject to copyright.  You can pretty much always learn from someone else's code and use similar ideas and patterns without concerns for license obligations, etc.  Sometimes the ideas in the code may be covered by patents, but that is seldom a concern for simple client libraries like go-github.

In short, yes please do copy the patterns!

**Comment by jon-whit** [comment]
@willnorris awesome, thank you very much for your detailed response! Happy coding :)
